src/util/walkController.ts

```diff
@@ -0,0 +1,132 @@
+  readonly taskExecutor: PrioritizedTaskExecutor
+  readonly trie: BaseTrie
+  private resolve: Function
+
+  /**
+   * Creates a new WalkController
+   * @param onNode - The `FoundNodeFunction` to call if a node is found
+   * @param trie - The `Trie` to walk on
+   */
+  private constructor(onNode: FoundNodeFunction, trie: BaseTrie, poolSize: number) {
+    this.onNode = onNode
+    this.taskExecutor = new PrioritizedTaskExecutor(poolSize)
+    this.trie = trie
+    this.resolve = () => {}
+  }
+
+  /**
+   * Async function to create and start a new walk over a trie.
+   * @param onNode - The `FoundNodeFunction to call if a node is found
+   * @param trie - The trie to walk on
+   * @param root - The root key to walk on
+   * @param poolSize - Task execution pool size to prevent OOM errors. Defaults to 500.
+   */
+  static async newWalk(
+    onNode: FoundNodeFunction,
+    trie: BaseTrie,
+    root: Buffer,
+    poolSize?: number
+  ): Promise<void> {
+    const strategy = new WalkController(onNode, trie, poolSize ?? 500)
+    await strategy.startWalk(root)
+  }
+
+  private async startWalk(root: Buffer): Promise<void> {
+    return await new Promise((resolve, reject) => {
+      this.resolve = resolve
+      this.trie
+        ._lookupNode(root)
+        .then((node) => {
+          this.processNode(root, node, [])
+        })
+        .catch((e) => {
+          reject(e)
+        })
+    })
+  }
+
+  /**
+   * Run all children of a node. Priority of these nodes are the key length of the children
+   * @param node - Node to get all children of and call onNode on
+   * @param key - The current `key` which would yield the `node` when trying to get this node with a `get` operation.
+   */
+  allChildren(node: TrieNode, key: Nibbles = []) {
+    if (node instanceof LeafNode) {
+      return
+    }
+    let children
+    if (node instanceof ExtensionNode) {
+      children = [[node.key, node.value]]
+    } else if (node instanceof BranchNode) {
+      children = node.getChildren().map((b) => [[b[0]], b[1]])
+    }
+    if (!children) {
+      return
+    }
+    for (const child of children) {
+      const keyExtension = child[0] as Nibbles
+      const childRef = child[1] as Buffer
+      const childKey = key.concat(keyExtension)
+      const priority = childKey.length
+      this.pushNode(childRef, childKey, priority)
+    }
+  }
+
+  /**
+   *
+   * @param nodeRef - Push a node reference to the event queue. This reference is a 32-byte keccak hash of the value corresponding to the `key`.
+   * @param key - The current key.
+   * @param priority - Optional priority, defaults to key length
+   */
+  pushNode(nodeRef: Buffer, key: Nibbles = [], priority?: number) {
+    this.taskExecutor.execute(priority ?? key.length, async (taskCallback: Function) => {
+      const childNode = await this.trie._lookupNode(nodeRef)
+      taskCallback()
+      this.processNode(nodeRef as Buffer, childNode as TrieNode, key)
+    })
```

Comment:
I have a bit hard time to read this. Is this executed (as suggested when reading `taskExecutor.execute()`) or just pushed to a queue like reading in the code documentation? Any way we can name this better? Is the existing `taskExecutor.execute()` named properly? 🤔

If it is more on the queuing side I would prefer a more explicit name like `pushNodeToQueue`, I remember that I always had my problems with reading/interpreting `pushNode()`, this time on review actually again.

Reply:
(comment is about the whole `pushNode()` function)

Reply:
You are right that the names are a bit ambiguous, I will edit those.